File: leetcode/leetcode_cpp/smallest-number-in-infinite-set.cpp
/*
https://leetcode.com/problems/smallest-number-in-infinite-set

1. Using set
time: o(logn)
space: o(n)
*/
class SmallestInfiniteSet {
public:
    SmallestInfiniteSet() {
        for (int i=1; i<=1000; ++i) s.insert(i);
    }
    
    int popSmallest() {
        int cur = *s.begin();
        s.erase(s.begin());
        return cur;
    }
    
    void addBack(int num) {
        s.insert(num);
    }
    set<int> s;
};

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet* obj = new SmallestInfiniteSet();
 * int param_1 = obj->popSmallest();
 * obj->addBack(num);
 */

File: leetcode/leetcode_py/smallest-number-in-infinite-set.py
'''
https://leetcode.com/problems/smallest-number-in-infinite-set

1. Using set
time: o(logn)
space: o(n)
'''
class SmallestInfiniteSet:

    def __init__(self):
        self.cur = 1
        self.s = set()
        
    def popSmallest(self) -> int:
        if self.s:
            res = min(self.s)
            self.s.remove(res)
            return res
        res, self.cur = self.cur, self.cur+1
        return res

    def addBack(self, num: int) -> None:
        if self.cur <= num:
            return
        self.s.add(num)
        


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)